tool/net/ltls.inc

```diff
@@ -0,0 +1,293 @@
+
+  if ((ret = mbedtls_ssl_set_hostname(&tls->ssl, server_name)) != 0) {
+    lua_pushnil(L);
+    lua_pushfstring(L, "mbedtls_ssl_set_hostname failed: %d", ret);
+    return 2;
+  }
+
+  mbedtls_ssl_set_bio(&tls->ssl, &tls->server_fd, mbedtls_net_send, NULL,
+                      mbedtls_net_recv_timeout);
+
+  if ((ret = mbedtls_ssl_handshake(&tls->ssl)) != 0) {
+    lua_pushnil(L);
+    lua_pushfstring(L, "SSL handshake failed: %d", ret);
+    return 2;
+  }
+
+  tls->connection_state = TLS_STATE_CONNECTED;
+
+  lua_pushboolean(L, 1);
+  return 1;
+}
+
+static int tls_write(lua_State *L) {
+  TlsContext **tlsp = checktls(L);
+  TlsContext *tls = *tlsp;
+  size_t len;
+  const char *data = luaL_checklstring(L, 2, &len);
+  int ret = mbedtls_ssl_write(&tls->ssl, (const unsigned char *)data, len);
+
+  if (ret < 0) {
+    lua_pushnil(L);
+    lua_pushfstring(L, "SSL write failed: %d", ret);
+    return 2;
+  }
+
+  lua_pushinteger(L, ret);
+  return 1;
+}
+
+static int tls_read(lua_State *L) {
+  TlsContext **tlsp = checktls(L);
+  TlsContext *tls = *tlsp;
+  lua_Integer bufsiz = luaL_optinteger(L, 2, BUFSIZ);
+  bufsiz = MIN(bufsiz, 0x7ffff000);
+
+  if (tls->read_buffer == NULL || tls->read_buffer_size < bufsiz) {
+    char *new_buffer = realloc(tls->read_buffer, bufsiz);
+    if (new_buffer == NULL) {
+      lua_pushnil(L);
+      lua_pushstring(L, "Memory allocation failed");
+      return 2;
+    }
+    tls->read_buffer = new_buffer;
+    tls->read_buffer_size = bufsiz;
+  }
+
+  int ret =
+      mbedtls_ssl_read(&tls->ssl, (unsigned char *)tls->read_buffer, bufsiz);
+
+  if (ret > 0) {
+    lua_pushlstring(L, tls->read_buffer, ret);
+    return 1;
+  } else if (ret == 0) {
+    // End of file
+    lua_pushnil(L);
+    return 1;
+  } else {
+    lua_pushnil(L);
+    if (ret == MBEDTLS_ERR_SSL_WANT_READ || ret == MBEDTLS_ERR_SSL_WANT_WRITE) {
+      lua_pushstring(L, "EAGAIN");
+    } else {
+      lua_pushfstring(L, "Read error: %d", ret);
+    }
+    return 2;
+  }
+}
+
+static int tls_close(lua_State *L) {
+  TlsContext **tlsp = checktls(L);
+  TlsContext *tls = *tlsp;
+
+  mbedtls_net_free(&tls->server_fd);
+  mbedtls_ssl_free(&tls->ssl);
+  mbedtls_ssl_config_free(&tls->conf);
+  mbedtls_ctr_drbg_free(&tls->ctr_drbg);
+  mbedtls_entropy_free(&tls->entropy);
+  free(tls->read_buffer);
+  tls->read_buffer = NULL;
+  tls->read_buffer_size = 0;
+  tls->connection_state = TLS_STATE_CLOSED;
+
+  return 0;
+}
+
+static int tls_tostring(lua_State *L) {
+  TlsContext **tlsp = checktls(L);
+  TlsContext *tls = *tlsp;
+  const char *state_str;
+
+  switch (tls->connection_state) {
+    case TLS_STATE_INIT:
+      state_str = "initialized";
+      break;
+    case TLS_STATE_CONNECTED:
+      state_str = "connected";
+      break;
+    case TLS_STATE_CLOSED:
+      state_str = "closed";
+      break;
+    default:
+      state_str = "unknown";
+  }
+
+  lua_pushfstring(L, "TLS connection (%p): %s", tls, state_str);
```

Comment:
With redbean what we like to do is have a `__repr` method so when an object is printed on the redbean shell, it'll print the code you'd type to create that object. Then the `__tostring` method will usually be the same thing as `__repr`.  I don't think it's worth showing state here, because the state you're showing is just the internal class state and might not reflect the current status in the OS of the socket.

Reply:
Ok. I will use:

``` C
static int tls_tostring(lua_State *L) {
  TlsContext **tlsp = checktls(L);
  TlsContext *tls = *tlsp;

  lua_pushfstring(L, "tls.TlsClient(fd=%d)", tls->fd);
  return 1;
}

static const struct luaL_Reg tls_methods[] = {
    {"write", tls_write},
    {"read", tls_read},
    {"close", tls_close},
    {"__gc", tls_gc},
    {"__tostring", tls_tostring},
    {"__repr", tls_tostring},
    {NULL, NULL}
};
```